specs/CriticalRestartRequired.md

```diff
@@ -0,0 +1,105 @@
+Critical Update Available
+===
+
+# Background
+As WebView2 developers, we often have to author ECS (experimentation and configuration service - 
+used to perform A/B testing or remotely disable features) configurations to toggle feature flags. 
+However, once these payloads are received, there is no way to restart the WebView2 and apply the
+payload. The purpose of this API is to detect such critical payloads and inform the end developer
+so they may restart their app or their WebView2 or other appropriate action.
+
+# Examples
+## WinRT and .NET   
+```c#
+void WebView_CoreWebView2InitializationCompleted(object sender, CoreWebView2InitializationCompletedEventArgs e)
+{
+    if (e.IsSuccess)
+    {
+        // ...
+        webView.CoreWebView2.Environment.CriticalRestartRequired += WebView_CriticalRestartRequired;
+    }
+}
+
+void WebView_CriticalRestartRequired(object sender, object e)
+{
+    // Depending on your app experience, you may or may not
+    // want to prompt user to restart the app.
+    RestartIfSelectedByUser();
+}
+```
+
+## Win32 C++
+```cpp
+void CoreWebView2InitializationCompleted() {
+    wil::com_ptr<ICoreWebView2Environment> m_webViewEnvironment;
+    auto env10 = m_webViewEnvironment.try_query<ICoreWebView2Environment10>();
+    CHECK_FAILURE(env10->add_CriticalRestartRequired(
+        Callback<ICoreWebView2CriticalRestartRequiredEventHandler>(
+            [this](ICoreWebView2Environment* sender, IUnknown* args) -> HRESULT
+            {
+                // Depending on your app experience, you may or may not
+                // want to prompt user to restart the app.
+                RestartIfSelectedByUser();
+                return S_OK;
+            })
+            .Get(),
+        nullptr));
+}
+```
+
+# API Notes
+
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+## Win32 C++
+
+```IDL
+interface ICoreWebView2Environment10;
+interface ICoreWebView2CriticalRestartRequiredEventHandler;
+
+[uuid(62cb67c6-b6a9-4209-8a12-72ca093b9547), object, pointer_default(unique)]
+interface ICoreWebView2CriticalRestartRequiredEventHandler : IUnknown {
+  /// Provides the event args for the corresponding event.  No event args exist
+  /// and the `args` parameter is set to `null`.
+  HRESULT Invoke([in] ICoreWebView2Environment* sender, [in] IUnknown* args);
+}
+
+[uuid(ef7632ec-d86e-46dd-9d59-e6ffb5c87878), object, pointer_default(unique)]
+interface ICoreWebView2Environment10 : IUnknown {
+  /// Add an event handler for the `CriticalRestartRequired` event.
```

Comment:
I can see us using this in the future for other types of restarts, not just critical ones.

* Reliability (turning off a feature that is buggy - not critical)
* Security (turning off a feature that has a security vulnerability - critical)
* Performance (turning off a feature that wastes memory - not critical)
* A/B testing (turning features on/off to assess whether they are worthwhile - not critical)

Maybe just "RestartRequested", and in the future we could add a Kind so the app can say "For security issues, I autosave and restart immediately, but for reliability and performance issues, I'll wait until the user finishes editing the current record, and for A/B testing, I'll wait until the user signs out of the app."

Reply:
That sounds good to me. Let's rename to 'RestartRequested'. In the future we can consider adding the rest.